<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Data input</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<!-- mouse_cortex_example.md is generated from mouse_cortex_example.Rmd Please edit that file -->

<pre><code class="r">library(Giotto)
</code></pre>

<h3>Data input</h3>

<ul>
<li>  load cortex/svz gene expression matrix</li>
<li>  prepare cell coordinates by stitching imaging fields</li>
</ul>

<p>Several fields - containing 100&#39;s of cells - in the mouse cortex and subventricular zone were imaged. The coordinates of the cells within each field are independent of eachother, so in order to visualize and process all cells together imaging fields will be stitched together by providing x and y-offset values specific to each field. These offset values are estimates based on the original raw image:
<img src="./cortex_svz_location_fields.png" alt="raw image" width="264" /> .</p>

<pre><code class="r">## visual cortex expression DATA ##
VC_exprs = read.table(system.file(&quot;extdata&quot;, &quot;cortex_svz_expression.txt&quot;, package = &quot;Giotto&quot;))

## prepare cell locations
VC_locs = fread(system.file(&quot;extdata&quot;, &quot;cortex_svz_centroids_rotated.csv&quot;, package = &quot;Giotto&quot;))
my_offset_file = data.table(field = c(0, 1, 2, 3, 4, 5, 6),
                            x_offset = c(0, 2048, 2048, 2048, 675, 2048, 675),
                            y_offset = c(2048, 2048, 2048, 2048,0, 0, 2048))
stitch_file = stitchFieldCoordinates(location_file = VC_locs, offset_file = my_offset_file,
                                     cumulate_offset_x = T, cumulate_offset_y = F,
                                     field_col = &#39;Field of View&#39;,
                                     reverse_final_x = F,
                                     reverse_final_y = T)
stitch_file    = stitch_file[,.(X_final, Y_final)]
my_offset_file = my_offset_file[,.(field, x_offset_final, y_offset_final)]
</code></pre>

<hr/>

<p> </p>

<h3>1. Create Giotto object &amp; process data</h3>

<pre><code class="r">## create
VC_test &lt;- createGiottoObject(raw_exprs = VC_exprs, spatial_locs = stitch_file, offset_file = my_offset_file)
## filter
VC_test &lt;- filterGiotto(gobject = VC_test,
                        expression_threshold = 1,
                        minimum_detected_genes = 10,
                        minimum_expression_in_cell = 10,
                        expression_values = c(&#39;raw&#39;),
                        verbose = T)
#&gt; Number of cells removed:  0  out of  913 
#&gt; Number of genes removed:  0  out of  10000
## normalize
VC_test &lt;- normalizeGiotto(gobject = VC_test)
# gene and cell statistics
VC_test &lt;- addStatistics(gobject = VC_test)
# adjust for covariates
VC_test = adjustGiottoMatrix(gobject = VC_test, expression_values = c(&#39;normalized&#39;),
                             batch_columns = NULL, covariate_columns = c(&#39;nr_genes&#39;, &#39;total_expr&#39;),
                             return_gobject = TRUE,
                             update_slot = c(&#39;custom&#39;))
# plain visualization
visPlot(gobject = VC_test)
#&gt; first and second dimenion need to be defined, default is first 2
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-4-1.png" width="75%" style="display: block; margin: auto;" /></p>

<hr/>

<p> </p>

<h3>2. dimension reduction</h3>

<pre><code class="r">## HVG genes
VC_test &lt;- calculateHVG(gobject = VC_test)
#&gt; 
#&gt;   no  yes 
#&gt; 8771 1229
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-5-1.png" width="50%" style="display: block; margin: auto;" /></p>

<pre><code class="r"># selected genes
gene_metadata = fDataDT(VC_test)
featgenes = gene_metadata[(hvg == &#39;yes&#39;) &amp; perc_cells &gt; 4 &amp; mean_expr_det &gt; 0.5]$gene_ID
# pca
VC_test &lt;- runPCA(gobject = VC_test, genes_to_use = featgenes)
# umap
VC_test &lt;- runUMAP(VC_test)
# tsne
VC_test &lt;- runtSNE(VC_test)
</code></pre>

<hr/>

<p> </p>

<h3>3. cluster</h3>

<pre><code class="r">## cluster
# SNN
VC_test &lt;- createNearestNetwork(gobject = VC_test)

# cluster on network
VC_test = doLeidenCluster(gobject = VC_test, resolution = 0.5,
                          python_path = &quot;/Users/rubendries/Bin/anaconda3/envs/py36/bin/python&quot;)
</code></pre>

<pre><code class="r">plotUMAP(gobject = VC_test, cell_color = &#39;pleiden_clus&#39;, point_size = 1.5,
        show_NN_network = T, edge_alpha = 0.1)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-8-1.png" width="60%" style="display: block; margin: auto;" /></p>

<hr/>

<p> </p>

<h3>4. co-visualize</h3>

<pre><code class="r"># expression and spatial
visSpatDimPlot(gobject = VC_test, cell_color = &#39;pleiden_clus&#39;, dim_point_size = 2, spatial_point_size = 2)
#&gt; first and second dimenion need to be defined, default is first 2
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-9-1.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r"># relationship between clusters
clusterheatmap &lt;- showClusterHeatmap(gobject = VC_test, cluster_column = &#39;pleiden_clus&#39;)
print(clusterheatmap)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-9-2.png" width="60%" style="display: block; margin: auto;" /></p>

<hr/>

<p> </p>

<h3>5. differential expression</h3>

<pre><code class="r"># pairwise t-test #
gene_markers = findMarkers(gobject = VC_test, cluster_column = &#39;pleiden_clus&#39;)
gene_markers_pair = findMarkers(gobject = VC_test, cluster_column = &#39;pleiden_clus&#39;,
                                group_1 = c(1,3), group_2 = c(2,4,5))

# Gini markers #
gini_markers = findGiniMarkers(gobject = VC_test, cluster_column = &#39;pleiden_clus&#39;)
gini_markers_DT = gini_markers[, head(.SD, 3), by = &#39;cluster&#39;]
myheat = plotHeatmap(gobject = VC_test, genes = gini_markers_DT$genes,
                     cluster_column = &#39;pleiden_clus&#39;)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-10-1.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">violinPlot(gobject = VC_test, genes = c(&#39;Nptxr&#39;, &#39;Cplx1&#39;,  &#39;Fgfr3&#39;, &#39;Cldn5&#39;, &#39;Cldn11&#39;, &#39;Igfbp5&#39;, &#39;Sox2&#39;, &#39;Thbs4&#39;, &#39;Clic6&#39;),
           cluster_column = &#39;pleiden_clus&#39;)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-11-1.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code>#&gt; first and second dimenion need to be defined, default is first 2
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-12-1.png" width="100%" /></p>

<hr/>

<p> </p>

<h3>6. spatial network + grid</h3>

<pre><code class="r">## spatial network
VC_test &lt;- createSpatialNetwork(gobject = VC_test, k = 3)
VC_test &lt;- createSpatialNetwork(gobject = VC_test, k = 100, maximum_distance = 200, minimum_k = 1, name = &#39;distance_network&#39;)
visPlot(gobject = VC_test, show_network = T, network_color = &#39;blue&#39;, point_size = 1)
#&gt; first and second dimenion need to be defined, default is first 2
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-13-1.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">## spatial grid
VC_test &lt;- createSpatialGrid(gobject = VC_test,
                             sdimx_stepsize = 500,
                             sdimy_stepsize = 500,
                             minimum_padding = 50 )
# spatial pattern genes
VC_test = detectSpatialPatterns(gobject = VC_test, dims_to_plot = 1)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-13-2.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code>#&gt; [1] &quot;Dim.1&quot;
#&gt; [1] &quot;Dim.2&quot;
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-13-3.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">## spatial genes
VC_test &lt;- calculateSpatialGenes(gobject = VC_test, min_N = 20)
spatial_gene_DT &lt;- calculateSpatialGenes(gobject = VC_test , method = &#39;kmeans&#39;, return_gobject = F)
# visualize
visGenePlot(gobject = VC_test,  genes = c(&#39;Enpp2&#39;, &#39;Shank1&#39;, &#39;Nptxr&#39;, &#39;Sox2&#39;),
            scale_alpha_with_expression = T)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-13-4.png" width="60%" style="display: block; margin: auto;" /></p>

<hr/>

<p> </p>

<h3>7. HMRF</h3>

<pre><code class="r"># select 500 spatial genes
gene_data = fDataDT(VC_test)
spatial_genes = gene_data[SV == &#39;yes&#39; | spg == &#39;yes&#39;]$gene_ID
set.seed(seed = 1234)
spatial_genes = spatial_genes[sample(x = 1:length(spatial_genes), size = 500)]

# run HMRF
HMRFtest = doHMRF(gobject = VC_test, expression_values = &#39;scaled&#39;,
                  spatial_genes = spatial_genes,
                  k = 10,
                  betas = c(40, 4, 3),
                  output_folder = &#39;/Volumes/Ruben_Seagate/Dropbox/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Data/package_testHMRF/&#39;,
                  python_path = &quot;/Users/rubendries/Bin/anaconda3/envs/py36/bin/pythonw&quot;)
#&gt; 
#&gt;  expression_matrix.txt already exists at this location, will be used again 
#&gt; 
#&gt;  spatial_genes.txt already exists at this location, will be used again 
#&gt; 
#&gt;  spatial_network.txt already exists at this location, will be used again 
#&gt; 
#&gt;  spatial_cell_locations.txt already exists at this location, will be used again

# view HMRF results for multiple tested betas
viewHMRFresults(gobject = VC_test,
                HMRFoutput = HMRFtest,
                k = 10, betas_to_view = c(44, 48), point_size = 2)
#&gt; [1] &quot;/Users/rubendries/Bin/anaconda3/envs/py36/bin/pythonw /Library/Frameworks/R.framework/Versions/3.5/Resources/library/Giotto/python/get_result2.py -r /Volumes/Ruben_Seagate/Dropbox/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Data/package_testHMRF//result.spatial.zscore -a test -k 10 -b 44&quot;
#&gt; first and second dimenion need to be defined, default is first 2
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-14-1.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code>#&gt; [1] &quot;/Users/rubendries/Bin/anaconda3/envs/py36/bin/pythonw /Library/Frameworks/R.framework/Versions/3.5/Resources/library/Giotto/python/get_result2.py -r /Volumes/Ruben_Seagate/Dropbox/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Data/package_testHMRF//result.spatial.zscore -a test -k 10 -b 48&quot;
#&gt; first and second dimenion need to be defined, default is first 2
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-14-2.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">
# add the HMRF results of interest
VC_test = addHMRF(gobject = VC_test,
                  HMRFoutput = HMRFtest,
                  k = 10, betas_to_add = c(48))
#&gt; [1] &quot;/Users/rubendries/Bin/anaconda3/envs/py36/bin/pythonw /Volumes/Ruben_Seagate/Dropbox/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Data/Qian_input_files//get_result2.py -r /Volumes/Ruben_Seagate/Dropbox/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Data/package_testHMRF//result.spatial.zscore -a test -k 10 -b 48&quot;

# co-visualize
visSpatDimPlot(gobject = VC_test, cell_color = &#39;hmrf_k.10_b.48&#39;, dim_point_size = 2, spatial_point_size = 2)
#&gt; first and second dimenion need to be defined, default is first 2
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-14-3.png" width="60%" style="display: block; margin: auto;" /></p>

<hr/>

<p> </p>

<h3>8. spatial analysis</h3>

<h5>Cell-cell preferential proximity</h5>

<p><img src="./cell_cell_neighbors.png" alt="cell-cell" width="453" /></p>

<pre><code class="r">## cell-cell interaction ##
## calculate and visualize cell-cell proximities
cell_proximities = cellProximityEnrichment(gobject = VC_test, cluster_column = &#39;cell_types&#39;)
cellProximityBarplot(CPscore = cell_proximities)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-15-1.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">cellProximityHeatmap(CPscore = cell_proximities, order_cell_types = T)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-15-2.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">
cellProximityVisPlot(gobject = VC_test, interaction_name = &#39;Astrocyte-Oligo&#39;,
                     cluster_column = &#39;cell_types&#39;,
                     cell_color = &#39;cell_types&#39;, show_network = T, network_color = &#39;blue&#39;)
#&gt; first and second dimenion need to be defined, default is first 2
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-15-3.png" width="60%" style="display: block; margin: auto;" /></p>

<p> </p>

<h5>1 gene enrichment for cell-cell interactions</h5>

<p><img src="./single_gene_enrichemt.png" alt="cell-cell" width="453" /></p>

<pre><code class="r">cell_int_gene_scores = getCellProximityGeneScores(gobject = VC_test, cluster_column = &#39;cell_types&#39;)
#&gt; start  Outer Neuron-Outer Neuron 
#&gt; start  Endothelial-Outer Neuron 
#&gt; start  Astrocyte-Outer Neuron 
#&gt; start  Inner Neuron-Outer Neuron 
#&gt; start  Astrocyte-Astrocyte 
#&gt; start  Astrocyte-Inner Neuron 
#&gt; start  NSC-Outer Neuron 
#&gt; start  Inner Neuron-Oligo 
#&gt; start  Oligo-Outer Neuron 
#&gt; start  Inner Neuron-NSC 
#&gt; start  Endothelial-Inner Neuron 
#&gt; start  Inner Neuron-Inner Neuron 
#&gt; start  Endothelial-Endothelial 
#&gt; start  Astrocyte-Endothelial 
#&gt; start  Astrocyte-Oligo 
#&gt; start  Endothelial-Oligo 
#&gt; start  Oligo-Oligo 
#&gt; start  NSC-NSC 
#&gt; start  Astrocyte-NSC 
#&gt; start  Choroid Plexus-Choroid Plexus 
#&gt; start  Endothelial-NSC 
#&gt; start  NSC-Oligo 
#&gt; start  Choroid Plexus-NSC 
#&gt; start  Choroid Plexus-Endothelial 
#&gt; start  Astrocyte-Choroid Plexus

# selection
setorder(cell_int_gene_scores, -diff_spat)
selection = cell_int_gene_scores[nr_1 &gt; 5 &amp; nr_2 &gt; 5, head(.SD, 1), by = interaction][1:2]

plotCellProximityGeneScores(CPGscores = cell_int_gene_scores,
                            selected_interactions = selection$interaction[1],
                            selected_genes = selection$genes[1])
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-16-1.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">
plotCellProximityGeneScores(CPGscores = cell_int_gene_scores,
                            selected_interactions = selection$interaction,
                            selected_genes = selection$genes[1],
                            detail_plot = T, facet.scales = &#39;fixed&#39;,
                            simple_plot = T,
                            simple_plot_facet = &#39;genes&#39;,
                            facet.ncol = 1, facet.nrow = 2)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-16-2.png" width="60%" style="display: block; margin: auto;" /></p>

<p> </p>

<h5>gene-gene enrichment for cell-cell interactions</h5>

<p>example: ligand - receptor combinations</p>

<p><img src="./double_gene_enrichment.png" alt="cell-cell" width="453" /></p>

<pre><code class="r">LR_data = fread(system.file(&quot;extdata&quot;, &quot;mouse_ligand_receptors.txt&quot;, package = &#39;Giotto&#39;))
ligands = LR_data$mouseLigand
receptors = LR_data$mouseReceptor

my_subset_interactions = c(&#39;Endothelial-NSC&#39;, &#39;Inner Neuron-NSC&#39;)
LR_VC = getGeneToGeneScores(CPGscore = cell_int_gene_scores,
                            selected_genes = NULL,
                            selected_cell_interactions = my_subset_interactions,
                            specific_genes_1 = ligands, specific_genes_2 = receptors)
#&gt; 
#&gt;  use specific gene-gene interactions 
#&gt; 
#&gt;  start specific gene-gene interactions

# select top 2
setorder(LR_VC, -diff_spat)
pair_selection = LR_VC[nr_1 &gt; 5 &amp; nr_2 &gt; 5, head(.SD, 1), by = interaction][1:2]

# detailed plot
plotCellProximityGeneToGeneScores(GTGscore = LR_VC, 
                                  selected_interactions = pair_selection$interaction,
                                  selected_gene_to_gene = pair_selection$gene_gene, detail_plot = T)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-17-1.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">
# simple plot per gene-gene
plotCellProximityGeneToGeneScores(GTGscore = LR_VC, 
                                  selected_interactions = pair_selection$interaction,
                                  selected_gene_to_gene = pair_selection$gene_gene,
                                  simple_plot = T,
                                  simple_plot_facet = &#39;genes&#39;)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-17-2.png" width="60%" style="display: block; margin: auto;" /></p>

<pre><code class="r">
# simple plot per cell-cell interaction
plotCellProximityGeneToGeneScores(GTGscore = LR_VC, 
                                  selected_interactions = pair_selection$interaction,
                                  selected_gene_to_gene = pair_selection$gene_gene,
                                  simple_plot = T,
                                  simple_plot_facet = &#39;interaction&#39;)
</code></pre>

<p><img src="man/figures/README-unnamed-chunk-17-3.png" width="60%" style="display: block; margin: auto;" /></p>

</body>

</html>
