
<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- This line is from RStudio -->

# Giotto

<!-- badges: start -->

[![License: GPL
v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
![Last Commit](https://badgen.net/github/last-commit/RubD/Giotto/suite)
<!-- badges: end -->

The Giotto Suite package provides tools to process, analyze and
visualize **single-cell and subcellular spatial expression** data. The
underlying framework is generalizable to virtually all currently
available spatial datasets. We recently demonstrated the general
applicability on 10 different datasets created by 9 different
state-of-the-art spatial technologies, including *in situ* hybridization
(seqFISH+, merFISH, osmFISH), sequencing (Slide-seq, Visium, STARmap)
and imaging-based multiplexing/proteomics (CyCIF, MIBI, CODEX). These
technologies differ in terms of resolution (single cell vs multiple
cells), spatial dimension (2D vs 3D), molecular modality (protein vs
RNA), and throughput (number of cells and genes). More information and
documentation about Giotto Suite can be found at
<http://giottosuite.com/index.html>.

<!-- <img src="inst/images/general_figs/overview_datasets.png" /> -->

## References

-   [Dries, R., Zhu, Q. et al.Â Giotto, a toolbox for integrative
    analysis and visualization of spatial expression data. bioRxiv
    701680 (2019).](https://www.biorxiv.org/content/10.1101/701680v2)
