---
output: html_document
---

## Description 
* Giotto provides a flexible framework for common single-cell processing steps such as:  
    + quality control
    + normalization
    + dimension reduction
    + clustering and cell type annotation
* To facilitate the analysis of recently emerging high-throughput, but lower-resolution spatial transcriptomic technologies, such as 10X Genomics Visium and Slide-seq, Giotto has 3 implemented algorithms for estimating the spatial enrichment of different cell types by integration of known gene signatures or single-cell RNAseq expression and annotation data.
* Spatial information is retained through the formation of a *spatial grid* and/or a *spatial proximity network*, which is used to:
    + identify spatial genes
    + extract continuous spatial-expression patterns
    + identify discrete spatial domains using HMRF
    + explore cell-type/cell-type spatial interaction enrichment or depletion
    + calculate spatially increased ligand-receptor expression in cells of interacting cell type pairs
    + find interaction changed genes (ICG): genes that change expression in one cell type due to interaction with a neighboring cell type
* Giotto provides a number of options to visualize both 2D and 3D data and the outcome of Giotto can be interactively explored using [**Giotto Viewer**](http://spatial.rc.fas.harvard.edu/giotto-viewer/), which allows you to overlay the obtained results with raw or additional images of the profiled tissue section(s). 

Make sure to check out the [**Datasets**](https://rubd.github.io/Giotto/articles/index.html) section to see examples of the Giotto workflow.

\  

## Workflow diagram 

<img src="../inst/images/general_figs/diagram_giotto_workflow.png" />

\  

## HOWTOs  

Howto's examples:  

##### Giotto tips & tricks 
- [How to add metadata and subset a Giotto object?](./metadata_and_subset_VC.html)
- [Different ways of subsetting Giotto results?](./inst/faqs/subset_giotto/subset_giotto.md)
- [How to create global instructions and show or save your created plots?](./inst/faqs/instructions_and_plotting/instructions_and_plotting.md)
- [Different ways to visualize your spatial data?](./inst/faqs/visualization_options/visualization_options.md)
- [How to test and store multiple parameters or analyses?](./inst/faqs/work_with_multiple_analyses/work_with_multiple_analyses.md)
- How to create a giotto object with your own spatial network/grid, dimensions reduction, ... ?  

##### Giotto analyses 
- How to identify highly variable genes?  
- Different ways to cluster data?
- Methods to identify differentially expressed genes?  
- How to perform cell-type or cell-process gene signature spatial enrichment analysis?
- Algorithmes to find spatial genes, patterns or domains?
- Effect of neighboring cells on gene expression?  
- ...

##### Giotto Analyzer and Viewer interaction  
- How to switch between Giotto Analyzer and Viewer?


\  

