<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial single-cell transcriptomics pipeline. • Giotto</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Spatial single-cell transcriptomics pipeline.">
<meta property="og:description" content="Pipeline to process, analyze and visualize spatial single-cell expression data.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="giotto" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#giotto" class="anchor"></a>Giotto</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>The Giotto package provides tools to process, analyze and visualize <strong>single-cell spatial expression</strong> data. The underlying framework is generalizable to virtually all currently available spatial datasets. We recently demonstrated the general applicability on 10 different datasets created by 9 different state-of-the-art spatial technologies, including <em>in situ</em> hybridization (seqFISH+, merFISH, osmFISH), sequencing (Slide-seq, Visium, STARmap) and imaging-based multiplexing/proteomics (CyCIF, MIBI, CODEX). These technologies differ in terms of resolution (single cell vs multiple cells), spatial dimension (2D vs 3D), molecular modality (protein vs RNA), and throughput (number of cells and genes).</p>
<div id="description" class="section level2">
<h2 class="hasAnchor">
<a href="#description" class="anchor"></a>Description</h2>
<ol>
<li>Giotto provides a flexible framework for common single-cell processing steps such as:</li>
</ol>
<!-- end list --><ul>
<li>quality control<br>
</li>
<li>normalization<br>
</li>
<li>dimension reduction<br>
</li>
<li>clustering and cell type annotation</li>
</ul>
<!-- end list --><ol>
<li><p>To facilitate the analysis of recently emerging high-throughput, but lower-resolution spatial transcriptomic technologies, such as 10X Genomics Visium and Slide-seq, Giotto has 3 implemented algorithms for estimating the spatial enrichment of different cell types by integration of known gene signatures or single-cell RNAseq expression and annotation data.</p></li>
<li><p>Spatial information is retained through the formation of a <em>spatial grid</em> and/or a <em>spatial proximity network</em>, which is used to:</p></li>
</ol>
<!-- end list --><ul>
<li>identify spatial genes<br>
</li>
<li>extract continuous spatial-expression patterns<br>
</li>
<li>identify discrete spatial domains using HMRF<br>
</li>
<li>explore cell-type/cell-type spatial interaction enrichment or depletion</li>
<li>calculate spatially increased ligand-receptor expression in cells of interacting cell type pairs</li>
<li>find interaction changed genes (ICG): genes that change expression in one cell type due to interaction with a neighboring cell type</li>
</ul>
<!-- end list --><ol>
<li>Giotto provides a number of options to visualize both 2D and 3D data and the outcome of Giotto can be interactivily explored using <a href="http://spatial.rc.fas.harvard.edu/giotto-viewer/"><strong>Giotto Viewer</strong></a>, which allows you to overlay the obtained results with raw or additional images of the profiled tissue section(s).</li>
</ol>
<p>Make sure to check out the <strong>Examples</strong> and <strong>FAQ</strong> section to learn everything about the Giotto workflow.</p>
<p> </p>
</div>
<div id="latest-news" class="section level2">
<h2 class="hasAnchor">
<a href="#latest-news" class="anchor"></a>Latest News</h2>
<ul>
<li>New example with 3D-like spatial data of the mouse hypothalamic preoptic region using <a href="https://science.sciencemag.org/content/362/6416/eaau5324">merFISH</a> (<strong>NEW</strong>)<br>
</li>
<li>New examples on mouse kidney and brain using the recently released <a href="https://www.10xgenomics.com/spatial-transcriptomics/">10X Visium datasets</a> (<strong>NEW</strong>)</li>
<li>Added tools to identify spatial enrichment based on cell-type (or other) specific gene signature lists (<strong>NEW</strong>)</li>
<li>Updated support for 3D spatial data<br>
</li>
<li>Added support for the use of global instructions and automatically saving your plots (<strong>NEW</strong>)</li>
<li>Implemented <a href="https://github.com/Teichlab/SpatialDE">SpatialDE</a> and <a href="https://github.com/edsgard/trendsceek">trendsceek</a>
</li>
<li>Add wrapper for differential expression with <a href="https://github.com/RGLab/MAST">MAST</a>
</li>
<li>New example with 3D spatial data <a href="https://science.sciencemag.org/content/361/6400/eaat5691">STARmap</a>
</li>
<li>New example with the highly sensitive data from <a href="https://www.nature.com/articles/s41592-018-0175-z">osmFISH</a>
</li>
<li>New example on the Cerebellum with the scalable data from <a href="https://science.sciencemag.org/content/363/6434/1463">Slideseq</a>
</li>
<li>New example on mouse olfactory bulb using immobilized primers on glass slides from <a href="https://science.sciencemag.org/content/353/6294/78">Spatial Transcriptomics</a>
</li>
<li>Updated seqFISH+ cortex example (<strong>NEW</strong>)</li>
<li>Updated STARmap cortex example (<strong>NEW</strong>)</li>
<li>New examples for Ab-based large-scale systems such as cyCIF, CODEX and MIBI will follow soon.</li>
</ul>
<p> </p>
</div>
<div id="requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h2>
<ul>
<li>R (&gt;= 3.5.1)</li>
<li>Python (&gt;= 3.0)</li>
<li>Windows, MacOS, Linux</li>
</ul>
<p> </p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>See FAQs for additional information.</p>
<div id="r-installation" class="section level4">
<h4 class="hasAnchor">
<a href="#r-installation" class="anchor"></a>R installation</h4>
<p>You can install (~1-5 mins) Giotto with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(devtools) <span class="co"># if not installed: install.packages('devtools') </span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(remotes)  <span class="co"># if not installed: install.packages('remotes')</span></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="co"># to install the latest version (development)</span></a>
<a class="sourceLine" id="cb1-4" title="4">remotes<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remotes/man/install_github.html">install_github</a></span>(<span class="st">"RubD/Giotto"</span>)</a>
<a class="sourceLine" id="cb1-5" title="5"></a>
<a class="sourceLine" id="cb1-6" title="6"><span class="co"># to install a specific release version (example)</span></a>
<a class="sourceLine" id="cb1-7" title="7">remotes<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remotes/man/install_github.html">install_github</a></span>(<span class="st">"RubD/Giotto@v0.1.3"</span>)</a></code></pre></div>
</div>
<div id="python-tools-optional" class="section level4">
<h4 class="hasAnchor">
<a href="#python-tools-optional" class="anchor"></a>Python tools (optional)</h4>
<p>This is necessary to run all available analyses, including Leiden / Louvain clustering and to build and use the interactive visualization tool. An alternative, but less flexible, R version for Louvain clustering is also available. It is advisable to install everything within a specific conda environment and specify the python path at the beginning with createGiottoInstructions() or in the R function itself when required.</p>
<p>Required python modules: pandas / igraph / networkx / leidenalg</p>
<p>pip installation one-liner:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" title="1"><span class="ex">pip3</span> install pandas python-igraph networkx python-louvain leidenalg</a></code></pre></div>
<p>If pip install does not work, try installing within a <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-with-commands">conda environment</a>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" title="1"><span class="ex">conda</span> install -c anaconda pandas</a>
<a class="sourceLine" id="cb3-2" title="2"><span class="ex">conda</span> install -c conda-forge python-igraph</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="ex">conda</span> install -c anaconda networkx</a>
<a class="sourceLine" id="cb3-4" title="4"><span class="ex">conda</span> install -c conda-forge python-louvain</a>
<a class="sourceLine" id="cb3-5" title="5"><span class="ex">conda</span> install -c conda-forge leidenalg</a></code></pre></div>
</div>
<div id="hmrf" class="section level4">
<h4 class="hasAnchor">
<a href="#hmrf" class="anchor"></a>HMRF</h4>
<p>See <a href="http://www.nature.com/articles/nbt.4260"><strong>HMRF</strong></a> installation <a href="http://spatial.rc.fas.harvard.edu/install.html">instructions</a>.</p>
</div>
<div id="giotto-viewer" class="section level4">
<h4 class="hasAnchor">
<a href="#giotto-viewer" class="anchor"></a>Giotto Viewer</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" title="1"><span class="ex">pip3</span> install --user jsbeautifier</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="ex">pip3</span> install --user giotto-viewer --no-cache <span class="co">#add --no-deps if do not wish to upgrade dependency</span></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="ex">pip3</span> install --user smfish-image-processing --no-cache <span class="co">#add --no-deps if do not wish to upgrade dependency</span></a></code></pre></div>
<p> </p>
</div>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<ul>
<li>see <a href="https://github.com/RubD/spatial-datasets" class="uri">https://github.com/RubD/spatial-datasets</a> to find raw and pre-processed input data and Giotto scripts (in progress).</li>
<li>typical run time range for the different datasets on a personal computer is around 10~45 mins.<br>
</li>
<li>click on the image and try them out yourself.</li>
</ul>
<p><a href="./inst/examples/mouse_cortex_svz/seqfish_cortex_Giotto_v0.1.4.html"><img src="./inst/images/cortex_image_summary.png" alt="Cortex"></a> <a href="./inst/examples/mouse_starmap_cortex/starmap_cortex_Giotto_v0.1.2_update.html"><img src="./inst/images/starmap_cortex_image_summary.png" alt="STARmap"></a></p>
<p><a href="./inst/examples/mouse_osmFISH_SS_cortex/osmFISH_SS_cortex_Giotto_v0.1.2.html"><img src="./inst/images/osmFISH_SS_cortex_image_summary.png" alt="osmFISH"></a> <a href="./inst/examples/mouse_SpatTx_OB/mouse_SpatTx_OB_Giotto_v0.1.2.html"><img src="./inst/images/SpatTx_OB_image_summary.png" alt="Spatial Transcriptomics"></a></p>
<p><a href="./inst/examples/mouse_visium_kidney/mouse_visium_kidney_Giotto_v0.1.3.html"><img src="./inst/images/visium_kidney_image_summary.png" alt="visium-kidney"></a> <a href="./inst/examples/mouse_visium_brain/mouse_visium_brain_Giotto_v0.1.3.html"><img src="./inst/images/visium_brain_image_summary.png" alt="visium-brain"></a></p>
<p><a href="./inst/examples/mouse_slideseq_cerebellum/slideseq_cerebellum_Giotto_v0.1.2.html"><img src="./inst/images/slideseq_cerebellum_image_summary.png" alt="slideseq"></a> <a href="./inst/examples/mouse_hypothalamic_preoptic_region/merfish_hypo_Giotto_v0.1.2_update.html"><img src="./inst/images/merFISH_hypothalam_image_summary.png" alt="merFISH"></a></p>
<p>[<img src="./inst/images/MIBI_tumor_image_summary.png" alt="MIBI-TOF">]</p>
<p> </p>
</div>
<div id="faq" class="section level2">
<h2 class="hasAnchor">
<a href="#faq" class="anchor"></a>FAQ</h2>
<p>Howto’s and faqs examples:</p>
<div id="installation-1" class="section level5">
<h5 class="hasAnchor">
<a href="#installation-1" class="anchor"></a>Installation</h5>
<ul>
<li><a href="./inst/faqs/giotto_installation/installation_issues.html">First time R package installation</a></li>
<li><a href="./inst/faqs/giotto_installation/installation_issues.html">Clang error on MacOS</a></li>
</ul>
</div>
<div id="data-availability" class="section level5">
<h5 class="hasAnchor">
<a href="#data-availability" class="anchor"></a>Data availability</h5>
<ul>
<li>Where to find seqFISH+ and other ready-to-use datasets?<br><strong>Checkout <a href="https://github.com/RubD/spatial-datasets" class="uri">https://github.com/RubD/spatial-datasets</a> to find already preprocessed datasets</strong>
</li>
</ul>
</div>
<div id="giotto-tips--tricks" class="section level5">
<h5 class="hasAnchor">
<a href="#giotto-tips--tricks" class="anchor"></a>Giotto tips &amp; tricks</h5>
<ul>
<li><a href="./inst/faqs/metadata_and_subset/metadata_and_subset_VC.html">How to add metadata and subset a Giotto object?</a></li>
<li><a href="./inst/faqs/subset_giotto/subset_giotto.html">Different ways of subsetting Giotto results?</a></li>
<li><a href="./inst/faqs/instructions_and_plotting/instructions_and_plotting.html">How to create global instructions and show or save your created plots?</a></li>
<li><a href="./inst/faqs/visualization_options/visualization_options.html">Different ways to visualize your spatial data?</a></li>
<li><a href="./inst/faqs/work_with_multiple_analyses/work_with_multiple_analyses.html">How to test and store multiple parameters or analyses?</a></li>
<li>How to create a giotto object with your own spatial network/grid, dimensions reduction, … ?</li>
</ul>
</div>
<div id="giotto-analyses" class="section level5">
<h5 class="hasAnchor">
<a href="#giotto-analyses" class="anchor"></a>Giotto analyses</h5>
<ul>
<li>How to identify highly variable genes?<br>
</li>
<li>Different ways to cluster data?</li>
<li>Methods to identify differentially expressed genes?<br>
</li>
<li>How to perform cell-type or cell-process gene signature spatial enrichment analysis?</li>
<li>Algorithmes to find spatial genes, patterns or domains?</li>
<li>Effect of neighboring cells on gene expression?<br>
</li>
<li>…</li>
</ul>
</div>
<div id="giotto-analyzer-and-viewer-interaction" class="section level5">
<h5 class="hasAnchor">
<a href="#giotto-analyzer-and-viewer-interaction" class="anchor"></a>Giotto Analyzer and Viewer interaction</h5>
<ul>
<li>How to switch between Giotto Analyzer and Viewer?</li>
</ul>
<p> </p>
</div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ul>
<li><p>Dries, R. et al. Giotto, a pipeline for integrative analysis and visualization of single-cell spatial transcriptomic data. bioRxiv 701680 (2019). <a href="doi:10.1101/701680" class="uri">doi:10.1101/701680</a> <a href="https://www.biorxiv.org/content/10.1101/701680v1">link</a></p></li>
<li><p>Eng, C.-H. L. et al. Transcriptome-scale super-resolved imaging in tissues by RNA seqFISH+. Nature 1 (2019). <a href="doi:10.1038/s41586-019-1049-y" class="uri">doi:10.1038/s41586-019-1049-y</a> <a href="https://www.nature.com/articles/s41586-019-1049-y">link</a></p></li>
<li><p>Zhu, Q., Shah, S., Dries, R., Cai, L. &amp; Yuan, G.-C. Identification of spatially associated subpopulations by combining scRNAseq and sequential fluorescence in situ hybridization data. Nature Biotechnology (2018). <a href="doi:10.1038/nbt.4260" class="uri">doi:10.1038/nbt.4260</a> <a href="https://www.nature.com/articles/nbt.4260">link</a></p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Ruben Dries <br><small class="roles"> Author, maintainer </small>  </li>
<li>Huipeng Li <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Ruben Dries, Huipeng Li.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
